doc/design/intra-process.md

```diff
@@ -0,0 +1,83 @@
+# Intra-process delivery
+
+Several requirements, each with a section.
+
+## Identifying local endpoints
+Endpoints on the same process have the same content on the first 8 octets of their GUID prefix. 
+
+### GUID_t refactor
+* Provide methods `is_builtin()`, `is_on_same_process_as(other_guid)` and `is_on_same_host_as(other_guid)`
+* Consider other improvements
+    * Split Guid.h into three headers (guid, prefix and entity_id)
+    * Converting EntityId_t to uint32_t
+    * Use a union on GuidPrefix_t to efficiently compare parts of it
+
+### Getting pointers to local endpoints
+* Add methods on RTPSDomain to return a pointer to reader/writer given its guid
+* Should be aware of builtin endpoints
+
+## Intra-process reception thread
+In order to separate the processing of new data on the subscribers from the user calls to write on the publisher, and simulate reception from the transport layer, a local reception thread should be added. It should work in a similar way to the AsyncWriterThread: readers will register their interest when they are informed of new local data. When the thread is awakened, it calls a `notify_new_changes` method on all the interested readers.
```

Comment:
I think if the user sets a listener, then the thread writing the sample should be the one calling the user's listener, having the same behavior as the reception threads. This should be well documented. If the user doesn't want this behavior, then he can use a waitset.